operators/src/adapters/raster_time.rs

```diff
@@ -87,34 +101,77 @@ where
+    F1::Stream: Stream<Item = Result<RasterTile2D<T1>>>,
+    F2::Stream: Stream<Item = Result<RasterTile2D<T2>>>,
+    F1::Output: Future<Output = Result<F1::Stream>>,
+    F2::Output: Future<Output = Result<F2::Stream>>,
 {
     type Item = Result<(RasterTile2D<T1>, RasterTile2D<T2>)>;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // The adapter aligns the two input time series by querying both sources simultaneously.
+        // All spatial tiles for the current time step are aligned, s.t. they are both valid
+        // for the same time (minimum of both). Then both queries are reset to start at the end
+        // of the previous time step.
+
         if self.is_terminated() {
             return Poll::Ready(None);
         }
 
         let RasterTimeAdapterProjection {
             source_a,
             source_b,
+            mut query_a_b_fut,
             query_rect,
-            time_end,
             num_spatial_tiles,
             current_spatial_tile,
             mut stream,
             ended,
         } = self.project();
 
-        let next = ready!(stream.as_mut().poll_next(cx));
+        // The `stream` is a `Zip` of the two input streams.
+        if stream.is_none() {
+            // the `stream` is not initialized and we need to query both sources and "await" their results
+            if query_a_b_fut.is_none() {
+                // the queries are not initialized and we need to query both sources and "await" their results
+                let fut1 = source_a.query(*query_rect);
+                let fut2 = source_b.query(*query_rect);
+                let join = future::join(fut1, fut2);
+
+                query_a_b_fut.set(Some(join));
+            }
+
+            if let Some(queries_fut) = query_a_b_fut.as_mut().as_pin_mut() {
+                // poll the queries, if they are not ready yet, return, otherwise take the result and reset the future
+                let queries = ready!(queries_fut.poll(cx));
+                query_a_b_fut.set(None);
+
+                match queries {
+                    (Ok(stream_a), Ok(stream_b)) => {
+                        // both sources produced an output, set the stream to be consumed
+                        stream.set(Some(stream_a.zip(stream_b)));
+                    }
+                    (Err(e), _) | (_, Err(e)) => {
+                        // at least one source failed, output error and end the stream
+                        *ended = true;
+                        return Poll::Ready(Some(Err(e)));
+                    }
+                }
+            }
+        }
+
+        // get the next pair of tiles from the stream. As the stream has to exist, this is safe.
+        let next = if let Some(stream) = stream.as_mut().as_pin_mut() {
+            ready!(stream.poll_next(cx))
+        } else {
+            panic!("stream is None")
+        };
 
         match next {
             Some((Ok(tile_a), Ok(tile_b))) => {
```

Comment:
If we calculated this (lines below) upfront in the constructor, couldn't we make it easier since we do not have the option?

Reply:
yes, but we can't compute it before, because we need the first tile to compute it

Reply:
Shouldn't we then use states, e.g. an enum instead of a struct?

Reply:
I introduced states but still we need to have a tile form the source to get calculate the number of tiles :(